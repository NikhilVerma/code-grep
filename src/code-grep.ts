/**
 * This file is a tool to "grep" for code by using Abstract syntax trees
 * It allows developers to do refactorings by searching for code patterns
 * which would be hard to express using regular expressions
 *
 * - Member expressions create({ something }) vs xxx.create({ something })
 * - Detect file contents (to use TS or Flow etc)
 * - Support block statements - { onOpen_: null }
 * - 'return' outside of function (2:12)
 * - function ___() {if (started) { return; }}
 */

import parseCode, { CodeMask } from './parsers/babel';
import traverse, { Node, NodePath, TraverseOptions } from '@babel/traverse';

export function codeSearch(
    code: string,
    masks: CodeMask[],
    onError?: (err: any) => void,
    fileName?: string
) {
    try {
        const ast = parseCode(code, fileName);
        const matches: NodePath<Node>[] = [];

        masks.forEach((expressionMask) => {
            traverse(
                ast,
                getVisitorFromMask(expressionMask, (nodePath) => {
                    matches.push(nodePath);
                })
            );
        });

        return matches;
    } catch (err) {
        if (onError) {
            onError(err);
        } else {
            console.error(err);
        }

        return [];
    }
}

function getVisitorFromMask(mask: CodeMask, onMatch: (nodePath: NodePath<Node>) => void) {
    const type = mask.type;

    if (!type) {
        console.error('No type found in mask', mask);
        throw new Error('No type found in mask');
    }

    return {
        [type](nodePath: NodePath<Node>) {
            if (matches(nodePath.node, mask)) {
                onMatch(nodePath);
            }
        },
    } as TraverseOptions;
}

function isObject(value: any): value is object {
    return Object.prototype.toString.call(value) === '[object Object]';
}

function isArray(value: any): value is any[] {
    return Object.prototype.toString.call(value) === '[object Array]';
}

function matches(value: Node, mask: any): boolean {
    if (Object.prototype.toString.call(mask) === '[object RegExp]') {
        return mask.test(value);
    } else if (isObject(mask)) {
        return (
            isObject(value) &&
            //@ts-ignore
            !Object.entries(mask).some(([key, maskValue]) => !matches(value[key], maskValue))
        );
    } else if (isArray(mask) && isArray(value)) {
        const clone = [...value];

        return mask.every((maskItem) => {
            return clone.some((valueItem, index) => {
                if (matches(valueItem, maskItem)) {
                    // This prevents re-matching with same items (e.g if user searched for <A/><A/>)
                    clone[index] = null;
                    return true;
                }

                return false;
            });
        });
    }

    return value === mask;
}
