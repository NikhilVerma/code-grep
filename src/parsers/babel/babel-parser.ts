import { parse, ParserPlugin } from '@babel/parser';

export function babelParser(code: string, fileName?: string) {
    //@todo - better TS detection
    const isTypescript = fileName?.endsWith('tsx') || fileName?.endsWith('ts');

    return parse(code, {
        allowImportExportEverywhere: true,
        plugins: [
            'asyncGenerators',
            'bigInt',
            'classProperties',
            'dynamicImport',
            'exportDefaultFrom',
            'functionBind',
            'functionSent',
            'importMeta',
            'jsx',
            'nullishCoalescingOperator',
            'numericSeparator',
            'objectRestSpread',
            'optionalCatchBinding',
            'optionalChaining',
            'throwExpressions',

            // Shaves a second off parsing for 2k+ files
            isTypescript ? 'typescript' : false,
            isTypescript
                ? [
                      'decorators',
                      {
                          decoratorsBeforeExport: true,
                      },
                  ]
                : false,

            // 'classPrivateMethods',
            // 'classPrivateProperties',
            // 'doExpressions',
            // 'estree',
            // 'exportNamespaceFrom',
            // 'pipelineOperator',
        ].filter(Boolean) as ParserPlugin[],
    });
}
