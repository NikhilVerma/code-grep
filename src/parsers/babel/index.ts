export { babelParser as default } from './babel-parser';
export { CodeMask, createCodeMask } from './create-mask';
