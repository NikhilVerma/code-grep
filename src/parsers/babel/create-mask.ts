import { babelParser } from './babel-parser';
import { Statement, Expression } from '@babel/types';
import { Node } from '@babel/traverse';

const SKIP_MASK_KEYS = ['start', 'end', 'loc', 'selfClosing', 'extra'];

export type CodeMask = ReturnType<typeof createCodeMask>[number];

export function createCodeMask(code: string) {
    return createMask(getParsedCode(code).program.body).map(expandMask).flat();
}

type NodeType = Node | Statement | Expression;

function expandMask(mask: Partial<NodeType>): Array<Partial<NodeType>> {
    // A lot of code is parsed as an expression statement while the intention of the user
    // is to search for the expression inside the statement
    if (mask.type === 'ExpressionStatement' && mask.expression) {
        return [mask, ...expandMask(mask.expression)];
    }

    // If a user enters "<A /><B />" as their pattern, it's not valid JSX
    // so we wrap it with fragments, but the user could be looking to match neighbouring elements
    if (mask.type === 'JSXFragment') {
        return [
            { type: 'JSXFragment', children: mask.children },
            { type: 'JSXElement', children: mask.children },
        ];
    }

    //@todo - Add support for AssignmentExpression (a = 5) > VariableDeclarator (const a = 5)

    return [mask];
}

function getParsedCode(code: string) {
    try {
        return babelParser(`(${code})`);
    } catch (err) {
        // Try to fix user mistakes

        // Invalid JSX
        if (err.message.includes('Adjacent JSX elements must be wrapped in an enclosing tag')) {
            return babelParser(`<>${code}</>`);
        }

        // Self own, don't convert this code into an expression
        if (
            err.message.includes('import can only be used') ||
            err.message.includes('Unexpected token')
        ) {
            return babelParser(code);
        }

        throw err;
    }
}

function createMask(node: NodeType[]): Partial<NodeType>[];
function createMask(node: NodeType): Partial<NodeType>;

function createMask(node: any): any {
    if (node instanceof Array) {
        return node.map(createMask);
    } else if (node instanceof Object) {
        const nodeKeys: Array<keyof typeof node> = Object.keys(node);
        const output: Partial<typeof node> = {};

        return nodeKeys.reduce((out, key) => {
            key = String(key);
            if (SKIP_MASK_KEYS.indexOf(key) === -1 && !isEmpty(node[key])) {
                out[String(key)] = createMask(node[key]);
            }
            return out;
        }, output);
    } else {
        return node;
    }
}

// eslint-disable-next-line complexity
function isEmpty(val: any) {
    return (
        val === undefined ||
        val === null ||
        val === '___' ||
        (val instanceof Array && val.length === 0) ||
        (val instanceof Object && Object.keys(val).length === 0)
    );
}
