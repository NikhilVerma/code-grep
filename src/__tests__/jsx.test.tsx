import { codeSearch } from '../code-grep';
import { createCodeMask } from '../parsers/babel';

test('jsx expressions', () => {
    expect(
        codeSearch(
            `
    function TestComponent() {
        return (<>
            <Component prop="test" />
            <Component />
            <Component onClick={evt => alert(evt)} prop="test" />
            //@todo - It should match ExpressionContainers as well
            <Component onClick={evt => alert(evt)} prop={"test"} />
            <Component prop="test">
                Some value
            </Component>
        </>);
    }
    `,
            createCodeMask('<Component prop="test" />')
        ).map((nodePath) => nodePath.toString())
    ).toMatchSnapshot();
});

test('neighboring elements', () => {
    expect(
        codeSearch(
            `
function TestComponent() {
    return (<>
        <C>
            <A />
            <B />
        </C>
        <B>
            <A>Hello</A>
            <B>World</B>
        </B>
        <A />
        <D />
        <A />
        <B />
    </>);
}
`,
            createCodeMask('<A /><B />')
        ).map((nodePath) => nodePath.toString())
    ).toMatchSnapshot();
});

test('same jsx elements', () => {
    expect(
        codeSearch(
            `
function TestComponent() {
    return (<>
        <A />
        <B />
        <A />
    </>);
}
`,
            createCodeMask('<A /><A />')
        ).map((nodePath) => nodePath.toString())
    ).toMatchSnapshot();

    expect(
        codeSearch(
            `
function TestComponent() {
    return (<>
        <A />
        <B />
    </>);
}
`,
            createCodeMask('<A /><A />')
        ).map((nodePath) => nodePath.toString())
    ).toMatchSnapshot();
});
