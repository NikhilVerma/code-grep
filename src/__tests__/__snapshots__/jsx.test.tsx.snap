// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsx expressions 1`] = `
Array [
  "<Component prop=\\"test\\" />",
  "<Component onClick={evt => alert(evt)} prop=\\"test\\" />",
  "<Component prop=\\"test\\">
                Some value
            </Component>",
]
`;

exports[`neighboring elements 1`] = `
Array [
  "<>
        <C>
            <A />
            <B />
        </C>
        <B>
            <A>Hello</A>
            <B>World</B>
        </B>
        <A />
        <D />
        <A />
        <B />
    </>",
  "<C>
            <A />
            <B />
        </C>",
  "<B>
            <A>Hello</A>
            <B>World</B>
        </B>",
]
`;

exports[`same jsx elements 1`] = `
Array [
  "<>
        <A />
        <B />
        <A />
    </>",
]
`;

exports[`same jsx elements 2`] = `Array []`;
